Cip68 structs encode & decode with field names, not serialization keys

rjharmon opened the pull request:
- serialization keys are specified in .toJsonSchema()
 - cast.toUplcData() and fromUplcData() work using the defined field names for their interface, while encoding using the encodingKeys
 - uses type information provided by the compiler

rjharmon committed:
Cip68 structs encode & decode with field names, not serialization keys

 - serialization keys are specified in .toJsonSchema()
 - cast.toUplcData() and fromUplcData() work using the defined field
   names for their interface, while encoding using the encodingKeys
 - uses type information provided by the compiler

rjharmon committed:
Shorten 'encodingKey' in type-schema struct to 'key'

 - uses encodingKey elsewhere, drawing a clear distinction between
   the type-schema fieldname and its more descriptive use elsewhere
   ... in the compiler, before forming that schema
   ... and in Cast codegen & serialize/deserialize, based on that schema
 - use 'sfStruct' shorthand for single-field struct in Cast `dataPath`
 - no need for "MapData" terminology in diagnostic errors

rjharmon commented:
See new diff, updated with short `key` in type-schema

christianschmitz commented:
Whilst you have been waiting for this to merge I've made some small changes, so you might need to rebase

rjharmon committed:
add "test trace OK" flags in dev-time test scripts

rjharmon committed:
cast unreachable schema-switch case, avoid type errors on <never>.format

rjharmon committed:
removed conditional-ConstrData wrapper for mStruct

An mStruct never needs an implicit ConstrDat wrapper; only what's
explicitly defined by the contract script author in its declarations.

 - a mStruct as a data is just MapData
 - an enum WrappedStruct { Cip68: {d: mStruct, ... } } is explicitly
   a ConstrData(variant-tag, MapData, ... )
 - adjusted tests to match
 - restored edge-case 'variant' switch for contract with Datum as a
   recognized variant type `main(datum: SomeEnum::Variant)`

rjharmon committed:
Merge remote-tracking branch 'origin/main' into map-struct-encoding

christianschmitz reviewed (approved):
Looks good
